### PR TITLE
UCT/CUDA/CUDA_COPY: cuda copy uct skeleton

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4851/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4971/ for details (Mellanox internal link).

File: src/uct/cuda/cuda_copy/cuda_copy_ep.c [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2017.  ALL RIGHTS RESERVED.
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.c]: style:  copyright should be 2017

File: src/uct/cuda/cuda_copy/cuda_copy_ep.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2001-2017.  ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.h]: s/2001-//

File: src/uct/cuda/cuda_copy/cuda_copy_ep.h [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+*/
+
+#ifndef UCT_SYSV_EP_H
+#define UCT_SYSV_EP_H
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.h]: should be UCT_CUDA_COPY_EP_H

File: src/uct/cuda/cuda_copy/cuda_copy_ep.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#include <ucs/type/class.h>
+
+
+typedef struct uct_cuda_copy_ep_addr {
+    int                ep_id;
+} uct_cuda_copy_ep_addr_t;
+
+typedef struct uct_cuda_copy_ep {
+    uct_base_ep_t      super;
+    struct uct_cuda_copy_ep *next;
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_ep.h]: style: align  structure fields

File: src/uct/cuda/cuda_copy/cuda_copy_iface.c [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+static int uct_cuda_copy_iface_is_reachable(const uct_iface_h iface, const uct_device_addr_t *dev_addr,
+                                       const uct_iface_addr_t *iface_addr)
+{
+    return 1;
+}
+
+static ucs_status_t uct_cuda_copy_iface_query(uct_iface_h iface,
+                                         uct_iface_attr_t *iface_attr)
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_iface.c]: style: align func params

File: src/uct/cuda/cuda_copy/cuda_copy_iface.c [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+static UCS_CLASS_INIT_FUNC(uct_cuda_copy_iface_t, uct_md_h md, uct_worker_h worker,
+                           const uct_iface_params_t *params,
+                           const uct_iface_config_t *tl_config)
+{
+    UCS_CLASS_CALL_SUPER_INIT(uct_base_iface_t, &uct_cuda_copy_iface_ops, md, worker,
+                              params, tl_config UCS_STATS_ARG(params->stats_root)
+                              UCS_STATS_ARG(UCT_CUDA_COPY_TL_NAME));
+
+    if (strcmp(params->mode.device.dev_name, UCT_CUDA_DEV_NAME) != 0) {
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_iface.c]: it is better to use strncmp()

File: src/uct/cuda/cuda_copy/cuda_copy_iface.c [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+UCS_CLASS_DEFINE(uct_cuda_copy_iface_t, uct_base_iface_t);
+UCS_CLASS_DEFINE_NEW_FUNC(uct_cuda_copy_iface_t, uct_iface_t, uct_md_h, uct_worker_h,
+                          const uct_iface_params_t*, const uct_iface_config_t*);
+static UCS_CLASS_DEFINE_DELETE_FUNC(uct_cuda_copy_iface_t, uct_iface_t);
+
+
+static ucs_status_t uct_cuda_copy_query_tl_resources(uct_md_h md,
+                                                uct_tl_resource_desc_t **resource_p,
+                                                unsigned *num_resources_p)
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_iface.c]: style: align function params

File: src/uct/cuda/cuda_copy/cuda_copy_md.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+typedef struct uct_cuda_copy_md {
+    struct uct_md super;   /**< Domain info */
+} uct_cuda_copy_md_t;
+
+/**
+ * gdr copy domain configuration.
+ */
+typedef struct uct_cuda_copy_md_config {
+    uct_md_config_t super;
+
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_md.h]: style: extra line

File: src/uct/cuda/cuda_copy/cuda_copy_md.c [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+static ucs_status_t uct_cuda_copy_mem_reg(uct_md_h md, void *address, size_t length,
+                                     unsigned flags, uct_mem_h *memh_p)
+{
+    ucs_status_t rc;
+    uct_mem_h * mem_hndl = NULL;
+    mem_hndl = ucs_malloc(sizeof(void *), "cuda_copy handle for test passing");
+    if (NULL == mem_hndl) {
+      ucs_error("Failed to allocate memory for gni_mem_handle_t");
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_md.c]: style: indentation

File: src/uct/cuda/cuda_copy/cuda_copy_iface.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2017.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCT_CUDA_IFACE_H
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_iface.h]: UCT_CUDA_COPY_IFACE_H ?

File: src/uct/cuda/cuda_copy/cuda_copy_iface.c [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+{
+    uct_tl_resource_desc_t *resource;
+
+    resource = ucs_calloc(1, sizeof(uct_tl_resource_desc_t), "resource desc");
+    if (NULL == resource) {
+      ucs_error("Failed to allocate memory");
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_iface.c]: style: indentation

File: src/uct/cuda/cuda_copy/cuda_copy_iface.h [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+#define UCT_CUDA_COPY_TL_NAME    "cuda_copy"
+#define UCT_CUDA_DEV_NAME   "cudacopy0"
+
+#define CUDA_FUNC(func)  ({                             \
+ucs_status_t _status = UCS_OK;                          \
+do {                                                    \
+    CUresult _result = (func);                          \
+    if (CUDA_SUCCESS != _result) {                      \
+        ucs_error("[%s:%d] cuda failed with %d \n",     \
+         __FILE__, __LINE__,_result);                   \
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_iface.h]: style: indent by (

File: src/uct/cuda/cuda_copy/cuda_copy_md.c [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+static ucs_status_t uct_cuda_copy_mkey_pack(uct_md_h md, uct_mem_h memh,
+                                      void *rkey_buffer)
+{
+    return UCS_OK;
+}
+
+static ucs_status_t uct_cuda_copy_rkey_unpack(uct_md_component_t *mdc,
+                                         const void *rkey_buffer, uct_rkey_t *rkey_p,
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_md.c]: style: indent by (

File: src/uct/cuda/cuda_copy/cuda_copy_md.c [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+                                         void **handle_p)
+{
+    *rkey_p   = 0xdeadbeef;
+    *handle_p = NULL;
+    return UCS_OK;
+}
+
+static ucs_status_t uct_cuda_copy_rkey_release(uct_md_component_t *mdc, uct_rkey_t rkey,
+                                          void *handle)
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_md.c]: style: indent by (

File: src/uct/cuda/cuda_copy/cuda_copy_md.c [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+{
+    return UCS_OK;
+}
+
+static ucs_status_t uct_cuda_copy_mem_reg(uct_md_h md, void *address, size_t length,
+                                     unsigned flags, uct_mem_h *memh_p)
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_md.c]: style: indent by (

File: src/uct/cuda/cuda_copy/cuda_copy_md.c [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+static ucs_status_t uct_cuda_copy_mem_dereg(uct_md_h md, uct_mem_h memh)
+{
+    ucs_free(memh);
+    return UCS_OK;
+}
+
+static ucs_status_t uct_cuda_copy_query_md_resources(uct_md_resource_desc_t **resources_p,
+                                                unsigned *num_resources_p)
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_md.c]: style: indentation

File: src/uct/cuda/cuda_copy/cuda_copy_md.c [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+{
+    return uct_single_md_resource(&uct_cuda_copy_md_component, resources_p, num_resources_p);
+}
+
+static ucs_status_t uct_cuda_copy_md_open(const char *md_name, const uct_md_config_t *md_config,
+                                     uct_md_h *md_p)
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_md.c]: style: indentation

**Comment by bureddy** [comment]
@alex-mikheev  please review.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5041/ for details (Mellanox internal link).

File: src/uct/cuda/cuda_copy/cuda_copy_md.h [hunk 18]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2017.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCT_CUDA_COPY_H
+#define UCT_CUDA_COPY_H
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_md.h]: it should be UCT_CUDA_COPY_MD_H

File: src/uct/cuda/cuda_copy/cuda_copy_iface.c [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+
+/* Forward declaration for the delete function */
+static void UCS_CLASS_DELETE_FUNC_NAME(uct_cuda_copy_iface_t)(uct_iface_t*);
+
+
+static ucs_status_t uct_cuda_copy_iface_get_address(uct_iface_h tl_iface,
+                                                    uct_iface_addr_t *iface_addr)
+{
+    int *cuda_copy_addr = (int*)iface_addr;
+    *cuda_copy_addr = 0;
```
Comment (alex-mikheev) [inline src/uct/cuda/cuda_copy/cuda_copy_iface.c]: style: add extra line between var declaration and code

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5125/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5156/ for details (Mellanox internal link).

**Pull request merged**